### PR TITLE
Update existing UBOs instead of replacing

### PR DESCRIPTION
**Pull request by mwilsnd**
Update UBO contents in-place if a suitable buffer is found already present.

### CODE REVIEW

**Review by sjg-wdw — COMMENTED** [review]
I don't know if it belongs in this one, but I think we need some method of queuing up UBO updates.

For Metal, or anywhere we're potentially rendering multiple frames at once, we need a way to queue up the copy into the buffer.  But perhaps we revisit this with Metal.

**Comment by mwilsnd** [comment]
> I think we need some method of queuing up UBO updates.

I can see a Metal version of the buffers holding a block of staging memory, duplicated to whatever maximum frames in flight we want to support. For the most part I think that can be a hidden implementation detail, we'd just include something like a dirty/needsUpload flag for the current frame being recorded and dispatch a deferred transfer from the staging area during the upload pass.

**Comment by sjg-wdw** [comment]
> > I think we need some method of queuing up UBO updates.
> 
> I can see a Metal version of the buffers holding a block of staging memory, duplicated to whatever maximum frames in flight we want to support. For the most part I think that can be a hidden implementation detail, we'd just include something like a dirty/needsUpload flag for the current frame being recorded and dispatch a deferred transfer from the staging area during the upload pass.

The way we did it last time was an explicit copy action on a Metal command buffer.  That way we didn't need to keep an array of buffers around.  But let's revisit when we get to Metal.

File: src/mbgl/gl/uniform_buffer_gl.cpp [hunk 1]
```diff
@@ -23,5 +24,12 @@ UniformBufferGL::~UniformBufferGL() {
     }
 }
 
+void UniformBufferGL::update(const void* data, std::size_t size_) {
+    assert(size == size_);
```
Comment (TimSylvester) [inline src/mbgl/gl/uniform_buffer_gl.cpp:28]: Is `size_>size` a memory corruption and potential crash?  Maybe we should return or limit the size in release builds.

**Review by TimSylvester — APPROVED** [review]

File: src/mbgl/gl/uniform_buffer_gl.cpp (hunk 1 above)
Comment (stefankarschti) [reply src/mbgl/gl/uniform_buffer_gl.cpp:28]: I agree: assert + log error? + return

**Review by stefankarschti — APPROVED** [review]

**Review by alexcristici — APPROVED** [review]

**Review by TimSylvester — APPROVED** [review]

**Pull request merged**